k2/python/k2/rnnt_loss.py

```diff
@@ -0,0 +1,995 @@
+    other purposes.
+
+    Args:
+      lm:
+        Language model part of un-normalized logprobs of symbols, to be added to
+        acoustic model part before normalizing.  Of shape:
+           [B][S+1][C]
+        where B is the batch size, S is the maximum sequence length of
+        the symbol sequence, possibly including the EOS symbol; and
+        C is size of the symbol vocabulary, including the termination/next-frame
+        symbol.
+        Conceptually, lm[b][s] is a vector of length [C] representing the
+        "language model" part of the un-normalized logprobs of symbols,
+        given all symbols *earlier than* s in the sequence.  The reason
+        we still need this for position S is that we may still be emitting
+        the termination/next-frame symbol at this point.
+      am:
+        Acoustic-model part of un-normalized logprobs of symbols, to be added
+        to language-model part before normalizing.  Of shape:
+           [B][T][C]
+        where B is the batch size, T is the maximum sequence length of
+        the acoustic sequences (in frames); and C is size of the symbol
+        vocabulary, including the termination/next-frame symbol.  It reflects
+        the "acoustic" part of the probability of any given symbol appearing
+        next on this frame.
+      symbols:
+        A LongTensor of shape [B][S], containing the symbols at each position
+        of the sequence, possibly including EOS
+      termination_symbol:
+        The identity of the termination symbol, must be in {0..C-1}
+      boundary:
+        a LongTensor of shape [B, 4] with elements interpreted as
+        [begin_symbol, begin_frame, end_symbol, end_frame] that is treated as
+        [0, 0, S, T]
+        if boundary is not supplied.
+        Most likely you will want begin_symbol and begin_frame to be zero.
+    Returns:
+        (px, py) (the names are quite arbitrary).
+           px: logprobs, of shape [B][S][T+1]
+           py: logprobs, of shape [B][S+1][T]
+        in the recursion:
+          p[b,0,0] = 0.0
+          p[b,s,t] = log_add(p[b,s-1,t] + px[b,s-1,t],
+                             p[b,s,t-1] + py[b,s,t-1])
+          .. where p[b][s][t] is the "joint score" of the pair of subsequences
+          of length s and t respectively.  px[b][s][t] represents the
+          probability of extending the subsequences of length (s,t) by one in
+          the s direction, given the particular symbol, and py[b][s][t]
+          represents the probability of extending the subsequences of length
+          (s,t) by one in the t direction,
+          i.e. of emitting the termination/next-frame symbol.
+
+          px[:,:,T] equals -infinity, meaning on the "one-past-the-last" frame
+          we cannot emit any symbols.  This is simply a way of incorporating
+          the probability of the termination symbol on the last frame.
+    """
+    assert (
+        lm.ndim == 3
```

Comment:
I would recommend separating these conjunctions out.
If one of the assertions fails, you don't know exactly which one failed.